### ISSUE TITLE
Questions on speeding up ViT

### ISSUE THREAD

**Issue by DianCh** [facebookresearch/xformers#626]
# ❓ Questions and Help
Hi! Thanks for this excellent repo! I wonder if I want to speed up my ViT without necessarily using sparse pattern, i.e., with vanilla dense attention:

1. Should I use `scaled_dot_product_attention`? Or should I manually call `memory_efficient_attention`?
2. When do I need the other operators like: `memory_efficient_attention_backward`, `memory_efficient_attention_forward`, `memory_efficient_attention_forward_requires_grad`?

I tried simply switching the attention computation in my original ViT model to `scaled_dot_product_attention` and didn't see speed-up with MAE training (the max memory did lower a little bit, though). Anything else I need to do?

Thanks again!

**Comment by danthe3rd** [comment]
Hello @DianCh 

You should use `xformers.ops.memory_efficient_attention` directly, and it will work for inference and training. This is what 99% of the users should use.
This is currently a bit confusing, but the other functions (`memory_efficient_attention_*`) are lower-level functions, that you should use if you are writing a custom backward pass for instance.

cc @fmassa the names are a bit confusing. we should at least update the doc

**Comment by DianCh** [comment]
Thank you @danthe3rd !

I tried `memory_efficient_attention` and this time the max memory allocation dropped a little again but the time per epoch is still pretty much the same.

vanilla implementation - max mem: 7563, Total time: 0:06:22 (0.1528 s / it)
`scaled_dot_product_attention` - max mem: 6517, Total time: 0:06:25 (0.1540 s / it)
`memory_efficient_attention` - max mem: 5618, Total time: 0:06:24 (0.1537 s / it)

The ViT takes in 224x224 images and downsamples by 16, which gives me 14x14=196 tokens; I'm using fp32. Is there a typical sequence length from which the speed-up starts to show for fp32? Are there other things I need to do on top of switching the attention computation?

**Comment by danthe3rd** [comment]
A few questions:
- which vit architecture are you using? (Or alternatively, what is the embedding size and number of heads)
- what is your batch size?
- what is your GPU?

**Comment by DianCh** [comment]
@danthe3rd I'm using ViT-B from the [official MAE repo](https://github.com/facebookresearch/mae/blob/main/models_mae.py). Mini batch size is 64, with 8 accumulation steps (that's why max mem appears quite small) and 8 GPUs, so 64x8x8=4096 total batch size. GPUs are A100.

**Comment by danthe3rd** [comment]
Ok that makes sense. You should consider using bf16 or f16 for training, this is the most efficient setting for mem-eff, and where the gap with pytorch implementation is the most important

**Comment by DianCh** [comment]
@danthe3rd I see. I came to xFormer from FlashAttention mainly because it supports f32 - now it looks like f32 isn't meant for speed-up, am I understanding right?
Does f16 usually incur performance drop? Are there any benchmarking of xFormer that show the model performance when trained with f32 vs. f16? Thanks!

**Comment by danthe3rd** [comment]
We have some performance measurements on f16 on the [README](https://github.com/facebookresearch/xformers#results). Should be exactly the same as for bf16.

> it looks like f32 isn't meant for speed-up, am I understanding right?

It's more like it wasn't a priority for us, as most researchers will use f16 or bf16 for training. It could be improved, but it's not high on our todo-list.

> I came to xFormer from FlashAttention mainly because it supports f32

xFormers has multiple kernels to run the attention, and will use the fastest available depending on your inputs. FlashAttention is also one of them, so when switching to bf16 you might very well run FlashAttention under the hood (most likely will use the triton backend for the forward, and FlashAttention for the BW).

> Does f16 usually incur performance drop

Do you mean in terms of accuracy? If it's done correctly, it shouldn't, but sometimes it's not trivial to get stable training in f16/bf16. If your model was originally trained in f16/bf16, it should be fine. Otherwise it might be risky indeed

**Comment by DianCh** [comment]
@danthe3rd Thanks a lot for the explanation! A last couple of questions:

> as most researchers will use f16 or bf16 for training

I've been mostly using f32 for my research and it's great to know that f/bf16 are good options too, since I'm mainly concerned about the model performance drop (I mean accuracy). Could you please point me to some research projects that's done in f/bf16?

> If it's done correctly, it shouldn't

May I ask what do you mean by "done correctly"? What typically do I need to look out for?

**Comment by danthe3rd** [comment]
For instance the DINO project used [fp16](https://github.com/facebookresearch/dino/blob/main/main_dino.py#L78) - that was running on V100 at the time. Because of the limited range of fp16, you might need to scale the activations to avoid going to 0 or inf - you should check the [pytorch doc](https://pytorch.org/docs/stable/amp.html) on the topic.

bf16 should be more stable (a100+) so it's generally better to use it over f16 if you have A100+ GPUs.
You might also want to start with autocast, to only run select operations in half precision - should give you some speedup already, while keeping most non-GEMM operations in f32 (at the cost of conversions).

**Comment by DianCh** [comment]
This is very helpful - thank you very much for the info!

**Comment by DianCh** [comment]
@danthe3rd Reopened for a couple follow-up question:

Is `memory_efficient_attention` doing the dispatching to different implementations under the hood? I noticed that in its docstring it says it implements the attention from paper [“Self-Attention Does Not Need O(n^2) Memory”](http://arxiv.org/abs/2112.05682)

What's the difference with `scale_product_attention` and when should I be using this one?

Thank you!

**Comment by danthe3rd** [comment]
`scale_dot_product_attention` will not use memory efficient attention at the moment, it will do something equivalent to the regular PyTorch code that you already have in your VIT.
You should use `memory_efficient_attention`

**Comment by DianCh** [comment]
@danthe3rd After double checking my [training script](https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/engine_pretrain.py#L47-L48) of ViT it turns out the model was already put under mixed precision like this:
```python
with torch.cuda.amp.autocast():
    loss, _, _ = model(samples, mask_ratio=args.mask_ratio)
```
In this case, do you know why using `memory_efficient_attention` didn't bring much speed-up to the training? Any suggestions on where I can check? Thanks!

**Comment by danthe3rd** [comment]
These are the results you can expect on A100 for some VIT models:
https://github.com/facebookresearch/xformers#results

Maybe try using fp16 without autocast, so you can be sure everything is in half precision. This should match the results in the graphs above when comparing with regular PyTorch in terms of efficiency already (regardless of the NANs or instabilities of training)
